src/libostree/ostree-repo-pull.c

```diff
@@ -1322,6 +1324,165 @@ commitstate_is_partial (OtPullData   *pull_data,
+
+          for (i = 0, n = g_variant_n_children (refs); i < n; i++)
+            {
+              const char *refname;
+              g_autoptr(GVariant) ref_v = g_variant_get_child_value (refs, i);
+
+              g_variant_get_child (ref_v, 0, "&s", &refname);
+
+              if (g_strcmp0 (refname, ref) == 0)
+                  return TRUE;
+            }
+        }
+    }
+
+  return FALSE;
+}
+
+static char *
+get_remote_collection_id (OstreeRepo  *repo,
+                          const gchar *remote_name)
+{
+  g_autofree gchar *remote_collection_id = NULL;
+
+  /* TODO: This happens for local pulls, I guess we should assume no
+   * collection id for this case?
+   */
+  if (remote_name == NULL)
+    return NULL;
+
+  if (!ostree_repo_get_remote_option (repo, remote_name, "collection-id", NULL,
+                                      &remote_collection_id, NULL) ||
+      remote_collection_id == NULL)
+    return NULL;
+
+  return g_steal_pointer (&remote_collection_id);
+}
+
+static gboolean
+verify_collection_ref (OtPullData  *pull_data,
+                       GVariant    *commit,
+                       GError     **error)
+{
+  g_autofree char *remote_collection_id;
+  remote_collection_id = get_remote_collection_id (pull_data->repo,
+                                                   pull_data->remote_name);
+  if (remote_collection_id == NULL)
+    return TRUE;
+
+  g_autoptr(GVariant) metadata = NULL;
+  metadata = g_variant_get_child_value (commit, 0);
+
+  gboolean unpacked;
+  const char *collection_id;
+  const char *ref;
+  unpacked = g_variant_lookup (metadata,
+                               "ostree.commit.collection-ref",
+                               "(&s&s)",
+                               &collection_id,
+                               &ref);
+
+  if (!unpacked)
+    return glnx_throw (error,
+                       "expected commit metadata to have collection "
+                       "ref information, found none");
+
+  if ((collection_id == NULL) ||
+      (collection_id[0] == '\0') ||
+      (ref == NULL) ||
+      (ref[0] == '\0'))
+    return glnx_throw (error,
+                       "bad collection ref information in commit metadata");
+
+  if (!g_str_equal (collection_id, remote_collection_id))
+    return glnx_throw (error,
+                       "commit has collection id %s in metadata, "
+                       "while the remote it came from has collection id %s",
+                       collection_id, remote_collection_id);
+
+  g_autoptr(OstreeCollectionRef) collection_ref;
+  collection_ref = ostree_collection_ref_new (collection_id, ref);
+  if (!g_hash_table_contains (pull_data->requested_refs_to_fetch, collection_ref))
+    return glnx_throw (error,
+                       "commit has collection id %s and ref %s, "
+                       "which was not requested",
+                       collection_id, ref);
+
+  /* TODO: not sure about this step, if the summary is meant to be unsigned…
```

Comment:
I’m not sure about this step either, but not because the `summary` is unsigned. I’m unsure what value it brings, given that the commit checksums we are pulling have been looked up from the `summary` file in the first place (at the start of the pull operation), so this is just verifying something we already know.